src/snowflake/cli/api/connections.py

```diff
@@ -0,0 +1,206 @@
+
+logger = logging.getLogger(__name__)
+
+schema_pattern = re.compile(r".+\..+")
+
+
+@dataclass
+class ConnectionContext:
+    # FIXME: can reduce duplication using config.ConnectionConfig
+    connection_name: Optional[str] = None
+    account: Optional[str] = None
+    database: Optional[str] = None
+    role: Optional[str] = None
+    schema: Optional[str] = None
+    user: Optional[str] = None
+    password: Optional[str] = field(default=None, repr=False)
+    authenticator: Optional[str] = None
+    private_key_file: Optional[str] = None
+    warehouse: Optional[str] = None
+    mfa_passcode: Optional[str] = None
+    enable_diag: Optional[bool] = False
+    diag_log_path: Optional[Path] = None
+    diag_allowlist_path: Optional[Path] = None
+    temporary_connection: bool = False
+    session_token: Optional[str] = None
+    master_token: Optional[str] = None
+    token_file_path: Optional[Path] = None
+
+    VALIDATED_FIELD_NAMES = ["schema"]
+
+    def clone(self) -> ConnectionContext:
+        return replace(self)
+
+    def update(self, **updates):
+        """
+        Given a dictionary of property (key, value) mappings, update properties
+        of this context object with equivalent names to the keys.
+
+        Raises KeyError if a non-property is specified as a key.
+        """
+        field_map = {field.name: field for field in fields(self)}
+        for (key, value) in updates.items():
+            # ensure key represents a property
+            if key not in field_map:
+                raise KeyError(f"{key} is not a field of {self.__class__.__name__}")
+            setattr(self, key, value)
+
+    def __repr__(self) -> str:
+        """Minimal repr where empty (i.e. None) values have their keys omitted."""
+        items = [
+            f"{k}={repr(v)}"
+            for (k, v) in self.as_nonempty_dict().items()
+            if v is not None
+        ]
+        return f"{self.__class__.__name__}({', '.join(items)})"
+
+    def __setattr__(self, prop, val):
+        """Runs registered validators before setting fields."""
+        if prop in self.VALIDATED_FIELD_NAMES:
+            validate = getattr(self, f"validate_{prop}")
+            validate(val)
+        super().__setattr__(prop, val)
+
+    def validate_schema(self, value: Optional[str]):
+        if (
+            value
+            and not (value.startswith('"') and value.endswith('"'))
+            # if schema is fully qualified name (db.schema)
+            and schema_pattern.match(value)
+        ):
+            raise InvalidSchemaError(value)
+
+    def validate_and_complete(self):
+        """
+        Ensure we can create a connection from this context.
+        """
+        if not self.temporary_connection and not self.connection_name:
+            self.connection_name = get_default_connection_name()
+
+    def as_nonempty_dict(self) -> dict:
+        return {k: v for (k, v) in asdict(self).items() if v is not None}
```

Comment:
nit, this function is sandwiched between two connection-related ones, can we move it down?